spring-integration-smb/src/main/java/org/springframework/integration/smb/session/SmbSession.java

```diff
@@ -480,9 +482,46 @@ public String getHostPort() {
 		return url.getHost() + ":" + url.getPort();
 	}
 
+	/**
+	 * Returns the contents of the specified SMB resource as an array of SmbFile filenames.
+	 * In case the remote resource does not exist, an empty array is returned.
+	 * @param _path path to a remote directory
+	 * @return array of SmbFile filenames
+	 * @throws IOException on error conditions returned by a CIFS server or if the remote resource is not a directory.
+	 */
 	@Override
-	public String[] listNames(String path) {
-		throw new UnsupportedOperationException("Not implemented yet");
+	public String[] listNames(String _path) throws IOException {
+		List<String> fileNames = new ArrayList<>();
+		try {
+			SmbFile smbDir = createSmbDirectoryObject(_path);
+			if (!smbDir.exists()) {
+				if (logger.isWarnEnabled()) {
+					logger.warn("Remote directory [" + _path + "] does not exist. Cannot list resources.");
+				}
+				return fileNames.toArray(new String[0]);
+			}
+			else if (!smbDir.isDirectory()) {
+				throw new IOException("Resource [" + _path + "] is not a directory. Cannot list resources.");
+			}
+
+			SmbFile[] files = smbDir.listFiles();
```

Comment:
Hm. Looks like there is just `String[] list ()`  API in that `SmbFile`.
Why don't we use it for optimization?